### DISCUSSION TITLE
Lock description

### DISCUSSION THREAD

**Discussion by ImagiNFT** [CounterpartyXCP/CIPs#135] [General]
This space is to discuss implementation of a lock description operation to ensure inmutability of assets

**Comment by davestaxcp** [comment]
In my experience and opinion:

Users, Artists, Collectors and Traders have been wanting this as an optional feature for a long time to ensure that their tokens 

- description can remain locked (indefinitely)
- new token issuance owners can issue subassets
- new token issuance owners could mint more supply (if token is unlocked)

**Comment by keyuno** [comment]
If it ain't broke, don't fix it.

In case someone desires to "lock" the description, they could  [burn ](https://www.reddit.com/r/counterparty_xcp/comments/6ze8bf/comment/dmy0f9s/)the ownership of the asset.

Besides, I'm highly skeptical of the underlying concept of "ensuring immutability", as it could be seen as a subtle attack on historical NFTs and the individuals who played pivotal roles in establishing the entire field. 

Does the integrity of the OLGA 1/1 NFT comes into question simply because its description isn't locked? And if the owner were to lock the description, would it be akin to [cleaning ](https://en.wikipedia.org/wiki/Coin_cleaning) valuable coins (a frowned-upon practice that can harm their value)?

OLGA stands as the genesis of what can be strictly defined as a 1/1 non-fungible token, marking a pivotal moment in crypto history. Its significance is unparalleled, in my opinion. To draw a comparison, in the art market, the Salvator Mundi piece fetched a staggering [$450.3 million](https://en.wikipedia.org/wiki/List_of_most_expensive_paintings), and it's not even considered the most historically significant painting. 

Would it be appropriate for me to diminish the significance of OLGA by suggesting that it's not entirely immutable compared to 1/1 NFTs with ownership burned? Should I argue that, in reality, the first 1/1 NFT with burned ownership is the true embodiment of "locked," as its description is immutable, thus deserving the title of the first Bitcoin NFT ever? Absolutely not. Any attempt to undermine the significance of OLGA would be disrespectful and, frankly, immoral.

We should approach any changes to Counterparty protocol with extreme caution and conservatism, only implementing them when absolutely necessary. NFTs with historical significance hold immense potential and can serve as the best store of value in the digital era.

Given the controversy surrounding asset divisibility reset, one must question the underlying intentions. Is it merely to accommodate asset owners who previously struggled to grasp the value of 1/1 NFTs and are now attempting to pass off their assets as NFTs? Needless to say, this endeavor is futile, as blockchain transparency reveals the truth. It's akin to burning the token supply and passing it off as genuine 1/1 NFT—a lost cause.

The truth is straightforward and sometimes needs to be explicitly stated:

1. Hal Finney theorized cryptographic  trading cards in [1993](https://github.com/cryptoanarchywiki/mailing-list-archive/blob/master/archive/1993/01/ee44616c1d030cb0722be6e3e5ff9c16e6535f48514cbb881f09b27884275c14/index.html).

2. JP Janssen created the first 1/1 NFT in [2014 ](https://jpjanssen.com/timeline/counterparty.html?asset=nft)with OLGA.

3.  In [2017](https://www.reddit.com/r/counterparty_xcp/comments/6gyp2b/masterkey_event_economy/), I generalized the concept of 1/1 NFTs as event assets, applying it to tokenize any conceivable information.

Yet, over the years, there has been a consistent tendency to underestimate the significance of numeric assets on Counterparty, often dismissing them as spam. Following this, the concept of NFTs has faced mockery, with dismissive remarks (right-click-save). Of course, assets on Bitcoin are timestamped. Once created, any subsequent assets representing the same idea are merely replicas, akin to printing a copy of the Mona Lisa and passing it off as authentic.

Presently, Bitcoin NFTs are gaining popularity, and alongside this rise, we've seen the emergence of false narratives, particularly concerning the concept of immutability (often misrepresenting the true meaning of provably-prunable). There have been deliberate attempts to obfuscate the distinction between the representation of tokens (on-chain meme) and the actual tokens.

It's time for people to simply accept the truth: 1/1 NFTs were established years ago and hold significant historical value, regardless of whether Wikipedia's NFT timeline or any other social media influencer's timeline has caught up to it. 

We have cryptographic proof on our side.

**Reply by B0BSmiths** [reply]
Resetting assets is nothing to do with asset owners who previously struggled to grasp the value of 1/1 NFTs.   That is a miss understanding of how the blockchain works.  

If you have a divisible asset you want to make indivisible or an indivisible you want to make divisible then resetting allows this and is the only way to achieve it, it does not allow you to pretend there is no blockchain history of an assets issuance and its supply and its divisibility and its description(s) as all of that is recorded on chain for ever immutably

Locking the description of Olga is not immoral and it does not bring anything into question - again its history is clear as it is on chain, it would become another feature of its character, another stroke of the artists brush should the artist ever decide to pick up the brush.  

Its a fact the Olga outputs that carried the image were not even a description but a broadcast and they were spent so it is prune able - to pretend otherwise is to deny reality - it cant be compared to a modern utxo resident nft as its not a utxo resident nft.  Sure it does not have a pointer like other early NFTs but its outputs were spent and were not part of the token.  It showed the way for sure, i think its a classic and a master piece and locking its description is neither here nor there, its would simply be another action added to its blockchain footprint that everyone can verify and no one could deny

**Reply by keyuno** [reply]
Can @Jpja, @loon3, and @jotapea provide public comments on these issues? I'm exhausted by the non-technical, nonsensical, low-level discussions that are either driven by bad faith actors or ignorance.

**Comment by pinnate-modo** [comment]
Strongly in favor of making descriptions optionally lockable. 

>In case someone desires to "lock" the description, they could [burn ](https://www.reddit.com/r/counterparty_xcp/comments/6ze8bf/comment/dmy0f9s/)the ownership of the asset.

Unless I am mistaken, burning the ownership would mean that no further quantity of the asset can be issued, which is not necessarily the intention of locking the description; e.g. if there is parseable data in the description field for an asset metaprotocol, which however is meant to integrate with "native" Counterparty assets.

>We should approach any changes to Counterparty protocol with extreme caution and conservatism, only implementing them when absolutely necessary. 

We should not make a virtue out of necessity: Counterparty's obsolescent protocol development over the years is not indicative of conservatism, but rather lack of attention. In the beginning, when Counterparty was very actively developed, protocol development may have been _careful_ but it was not _conservative_ (e.g. rock-paper-scissors was implemented as a proof-of-concept for trustless gaming).

Counterparty wasn't perfect when it was first released and it isn't perfect today. As a protocol with hard-coded smart contracts, it is especially important that it be responsive to the community's needs.

Finally, I do not see how adding a boolean to issuances can be taken as an attack on historical NFTs or an attempt to denigrate earlier work; certainly nothing in the OP suggested that.

**Reply by keyuno** [reply]
> In the beginning, when Counterparty was very actively developed, protocol development may have been careful but it was not conservative (e.g. rock-paper-scissors was implemented as a proof-of-concept for trustless gaming).

With due respect, many of the initially envisioned use cases for Counterparty failed to materialize. The platform's enduring relevance largely hinged on collectibles, despite some erroneously dismissing them as easily replicable—a common misconception. Collectibles emerged as the lifeline for Counterparty, demonstrating their singular success as a use case.

Despite certain individuals downplaying the value of crypto-collectibles, they have proven indispensable. Those who overlook this fact continue to misunderstand the platform's true potential. Therefore, moving forward, Counterparty should adopt a conservative approach, eschewing the 'move fast, break things' mentality, especially considering the historically significant nature of its assets.

**Comment by adamkrellenstein** [comment]
This is easy to support: it's an optional feature that's simple to implement and has obvious user demand.

**Reply by jdogresorg** [reply]
Agreed, this feature has user demand and is pretty easy to implement. No objections here :)

**Comment by jotapea** [comment]
In my early days of getting involved and learning about the protocol, I thought the good old 'LOCK' description would lock the quantity and the description. This was incorrect, and nowadays lock is its own separate issuance parameter. Saying this because I think what would have made the most "intuitive" sense (at least for me), was to just make a 'LOCK' description lock both.

The user with issuance privileges, as long as they don't get transferred, could still issue more description "updates". These would not be seen as valid by the protocol... but still there. Which is what can still happen with a new lock-description. Non-privileged issuers can also do this, but is not fooling anyone as the protocol will also detect the different source. 

So, after reading @keyuno 's perspective of "If it ain't broke, don't fix it.", I kind of lean more into not adding more variables to something that is not really doing much in terms of true intentions (what the timechain shows). Locking quantities are critical for the ledger, but descriptions really aren't. Quantities are the main responsibility of the protocol, I believe.

The undeniable full-intent final "immutable" locking is burn address issuance transfer.

**Reply by jdogresorg** [reply]
Sending asset ownership to a burn address is a “hack” to lock an asset description. This “hack” results in being unable to issue additional subassets.

Generally I disagree with adding a lock description flag, however in this case, there is clear user demand for this functionality. I prefer to add a clear option to lock a description rather than use other methods with unintended consequences, like inability to do additional issuances, subasset issuances, etc.

**Reply by keyuno** [reply]
The push for this change seems to stem from two main reasons. First, there's the 'immutability' crowd, which promotes misleading ideas about 'on-chain' representations and "prunable" nonsense that, in my opinion, only appeals to speculators trying to undermine the value of historical NFTs for financial gain. Second, there's the concern that an asset owner might alter the description to something inappropriate. However, even with subassets remaining tied to the main asset, an owner could still issue something questionable, tarnishing the associated tokens.

Introducing changes to the asset structure without clear necessity and understanding of the market repercussions can lead to unintended harm. From a collector's perspective, the current asset structure isn't flawed, and introducing additional elements merely adds unnecessary risk—a slippery slope without clear benefits, making it an ethical misstep.

**Reply by davestaxcp** [reply]
I disagree Keyuno,

From a collector's (and artist) perspective, **the current asset structure is flawed** because it gives NO possibility to lock a description AND to do additional issuances, subasset issuances, etc.

because of the current structure, JPJA with his new CIP33 (https://jpja.github.io/Electrum-Counterparty/cip33_issuance.html) forces the token ownership to be sent to a burn address (so it is forever unchanged and immutible)

collectors and artists have been asking for this for YEARS..... 

and if my personal opinion matters here.... I also disagree with Jotapea and think the Token Issuance Lock and the Description Lock should be two separate functions as it opens more possibilities for artists to change things before locking the art (if it is apart of their art piece and collectors are fine with that)

**Reply by jotapea** [reply]
To be clear, I don't want to change the lock to also apply to the description. My point is that we should not add more complexity, and I would just leave it as it is...

OR maybe is better to do a new issuances design that considers this, and additional ideas like removing reset, sub-sub-assets, and maybe even viral "mints" option?

The way the protocol works, as of today, is that invalid issuances are still detected and stored. So, it doesn't stop a project to continue issuing description updates by the same source even after a "description-lock" is applied. And as the "reset" showed us, something that used to be "immutable", can me made "mutable" by a software upgrade. On the other hand, burn addresses are truly final, thus a superior "immutabilty" method.

And I get that the sub-asset issuances are lost by burning, but numeric assets cannot issue sub-assets anyway. So this proposal is really only useful for named (super)assets. And the point of allowing separate independent locks is making it even more complex which is not the best move IMO, ordinals don't even have a distinction between issuer and holder.

**Comment by B0BSmiths** [comment]
Counterparty was launched after burning BTC which was considered a hack by some, everything is a hack to an extent - we are putting JPEGs into the UTXO set using fake pubkeys, it would be nice to not have to use hacks on top of hacks and to have more flexibility when it comes to issuance.

My thinking was for trading asset ownerships.  There is a OG Rare that has a 0 issuance whose ownership is traded, so that poses the question should a subsequent owner of an asset be able to issue a new/additional description?  I am not talking about the person that simply possesses the token but the person that owns it.  

Should a subsequent owner be able to issue a new description on a token they never created but now own? if it was description locked it would prevent a subsequent owner from issuing a new/additional description updates

Some simple API sanity checks can be used to prevent owners from issuing updates after description is locked 

I have created some unlocked ownership retained Olga style stamps in the name of science.

I say give the artist more flexibility and the power to set the token as they wish - no one is forcing anyone to description lock anything - but to give the token creator the choice before its ownership trades hands

**Comment by jotapea** [comment]
Maybe... a better approach would be a new "final-issuance" type? Meaning that whatever you write in the genesis issuance is final. No quantity or description updates allowed.

This would make it more aligned to how the competitors are doing it.

Thinking we could have multiple (mutually exclusive) genesis issuance types:

- "name-only-issuances": allows users to issue an asset without having to specify anything else (description, quantity or divisibility). The most flexible as a future issuer can specify these.
- "final-issuances": as explained above. And the least flexible.
- "classic-issuances": like they work today minus the reset (name-only becomes the "reset" use-case).

Then, we could also add:

- "final-open-issuances": work just like final, but the issuer doesn't receive the complete supply. The supply is up for grabs by the subsequent issuers. Just like the open mints do. (This can even be considered as the ONLY new type, as specifying the complete supply as the minting quantity is effectively the same as the "final-issuances" explained above.)

Basically the UX would only need the specify the new final(open), because name-only can be implicit as long as the issuance history has never specified a quantity or description (and lock). So, some classic issuances can already be name-only.

And, I believe, this won't affect @keyuno 's arguments, as these final(open) issuances can be considered a whole new type of asset that lives side-by-side with the classic ones. Classic ones cannot become final, and vice-versa.

(Edited for a more complete argument.)

**Reply by keyuno** [reply]
> whole new type of asset that lives side-by-side with the classic ones

The problem lies in disrupting the structure of already established assets without a valid reason. Introducing an entirely new type of asset structure is beyond the scope of this discussion.

**Reply by jotapea** [reply]
Is not beyond the scope of this discussion, as the proposal of a new "genesis-final" option solves the lock-description argument for those that want it.

Basically, instead of adding more complexity to the current structure, we could simplify the current one (by removing reset), and then add something new that covers multiple use-cases not currently supported by the platform. And these are mutually exclusive, BUT sharing the same namespace. (So, maybe having a different burned-XCP requirement will also be appropriate.)

**Comment by Jpja** [comment]
> This space is to discuss implementation of a lock description operation to ensure inmutability of assets

First, I disagree with the premise "to ensure inmutability". Every action on Counterparty is embedded in a Bitcoin transaction, thus immutable. The asset description is no different – once set it can never be changed or removed, only appended with new one(s).

Immutability aside, the ability to lock an asset description has one clear benefit: General explorers (and the api) look at the last valid issuance. Special explorers, like Rare Pepe and Stamps, have their own meta-rules that might ignore issuances. Rare Pepe relies on an off-chain directory and Stamps looks at the _first_ issuance with a `STAMP:` prefix.

Locking of description will imo be a great feature for **general** use of Counterparty, and a cleaner solution compared with burn (although burn is more fun, aligned with the Counterparty ethos). However, burn addresses can be [tricky to detect](https://github.com/Jpja/XCP-Asset-Timeline/blob/main/js/burn_address.js) algorithmically and – as pointed out – appear hacky. 

I have personally wanted lock descriptions since 2015 and I know many others have asked for it too. If Counterparty started up today, I definitively would have supported  it.

My concerns now are due to economic implications. There is a billion dollar market for Counterparty **tokens**. Very few collect Counterparty **asset ownerships** although the CoinDaddy marketplace has enabled it since 2015. The asset ownership is more like a domain or contract ownership. A locked asset will become more like a token. What happens then when you have a rare token, supply 100, and one "super token" that is the locked asset ownership? Or what about PEPENOPOLOUS with supply 1, where the token was sold for $3.6 million .. is the asset ownership even more valuable?

I don't know the answer. Imo the token is the token and the asset ownership is the tool used to mint the token. The market will value these two collectibles accordingly. Perhaps a locked ownership is even **better** for the valuation of the token since the description is guaranteed never to change?

Another factor is utxo-binding for PSBT trading. This will likely enable risk free trading of asset ownerships too. This blurs the line between asset ownerships and tokens even more.

Finally some technical notes. Description lock will require a new flag, thus we need to allocate two new message IDs (asset and subasset issuances). There are only 128 of one-byte IDs before we need to move on to 4 bytes. Not a biggie as I believe only 30 or so are used, but good to keep in mind. Currently one byte is used per flag but with the new IDs I suggest one bit per flag, and allocating two bytes for the flags. Then we can add more flags later (eg lock subassets) without using up more IDs.

A second technical consideration is that one new line is written to the issuances table each time an asset ownership is transferred. This table is not indexed afaik, so people sending ownerships around like tokens might cause performance issues. My understanding is that this can be easily fixed though.

The bottom line is that this is more complex than it seems at first. I do, however, lean in **favor of implementing a lock description operation** provided collectors overwhelmingly are in favor of it (after hearing the arguments).

**Reply by jotapea** [reply]
> First, I disagree with the premise "to ensure inmutability". Every action on Counterparty is embedded in a Bitcoin transaction, thus immutable. The asset description is no different – once set it can never be changed or removed, only appended with new one(s).

100% agree with this.

**Reply by adamkrellenstein** [reply]
Assets on Counterparty are not immutable. Nor are balances, dispensers, orders, etc. 

Something is immutable when it doesn't change. Assets change, e.g. when their descriptions change. What're immutable are the transactions that trigger the changes in Counterparty: a _transaction_ causing a change in a description is definitely immutable (and the entries in the DB, etc. etc.), but the description itself is not. Be careful not to confuse the two.

As above, this would be a super simple fix to an obvious oversight in the original design. Everything else is hacky and much more complex.

**Reply by jotapea** [reply]
Is not about confusion, is about interpretation. A project can decide to ignore subsequent description updates (like stamps do). Also, a project can consider all descriptions history (bitSTART).

The protocol allowing both is a positive for creativity.

The reference implementation api selecting the last description is an "implementation detail" of the api itself. The api could also return the descriptions history (or just complete issuances history, which will cover the descriptions).

**Reply by keyuno** [reply]
> First, I disagree with the premise "to ensure inmutability". 

This proposed change is being used to promote a misleading narrative under the guise of 'ensuring immutability.' In my opinion, this tactic is intended to devalue historic NFTs. There is a precedent for such dirty tactics, as we have already seen a narrative attack on the 'immutability' of OP_RETURN.

> and a cleaner solution compared with burn

There is no genuine technical necessity to disrupt the established asset structure. What is considered elegant is subjective; some might view proof-of-burn as the elegant solution because it doesn't disrupt the existing asset structure.

> My concerns now are due to economic implications. 

Precisely. We do not know the market implications, and that’s the point. We don’t know how it will impact the collectors' market. Some engineers want to engineer for the sake of it, even when there is no actual need. We must keep things simple. If it ain't broke, don't fix it.

**Reply by B0BSmiths** [reply]
What is on chain is immutable.  UTXO set residence is a meme some enjoy.  I was coming at this description lock idea from the same angle that lead me to the burn keys idea, satoshis that can be spent will be spent, it was the spending of the Olga broadcast that showed it happens, and is why I thought a description lock could be a good idea.
